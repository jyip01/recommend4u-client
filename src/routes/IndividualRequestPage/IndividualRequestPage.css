.indivRequestPage{
  display:flex;
  height:76vh;
}

.requestInfo{
  background-color: #f5cdb4;
  border-radius:6px;
  padding: 10px;
  margin-bottom:25px;
}

.mainBar{
  padding-bottom:20px;
  overflow:scroll;
}

.comment{
  margin-left:30px;
  margin-top:10px;
  border:1px solid grey;
}
  
.requestData{
  display:flex;
  margin-top:10px;
  margin-bottom:10px;
}

.iconContainer{
  display:flex;
  flex-direction:column;
  line-height:20px;
}

.iconContainer i{
  line-height:22px;
}

.infoContainer{
  margin-left:10px;
}

.infoContainer p{
  line-height:22px;
}

.deleteButton{
  width:80px;
  background-color:#384a62;
  color:white;
  display:flex;
  justify-content: center;
  align-items:center;
  padding:5px;
  text-decoration: none;
  font-weight: bold;
  border-radius: 2px;
  margin-top:10px;
  border:0px;
}

.deleteButton:hover{
  background-color: #7496a2;
  cursor: pointer;
}

.editAndDeleteBox{
  display:flex;
  justify-content: space-between;
}

.editAndDeleteBox button{
  width:48%;
}

.commentsCount{
  padding-left:10px;
  margin:20px;
  font-size: 18px;
  font-weight: bold;
  text-decoration: underline #384a62;
}

.editButton{
  margin-top:10px;
  width:140px;
  border:1px solid #384a62;
  color:#384a62;
  display:flex;
  justify-content: center;
  align-items:center;
  padding:5px;
  text-decoration: none;
  font-weight: bold;
  border-radius: 2px;
  background-color:#f5d8c6;
}

.editButton:hover{
  background-color: #384a62;
  color:white;
  cursor: pointer;
}


