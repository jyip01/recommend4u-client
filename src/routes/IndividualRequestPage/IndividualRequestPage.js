import React, { Component } from 'react'
import './IndividualRequestPage.css'
import RequestApiService from '../../services/request-api-services'
import RequestListContext from '../../contexts/RequestListContext'
import CommentForm from '../../components/CommentForm/CommentForm'
import CommentItem from '../../components/CommentItem/CommentItem'
import UserService from '../../services/user-service'
import {Redirect} from 'react-router-dom'
import UpdateRequest from '../../components/UpdateRequest/UpdateRequest'
import Error from '../../components/Error/Error'

export default class IndividualRequestPage extends Component {
    
    static contextType = RequestListContext

    state = {
        commentAdded: false,
        toRequests: false,
        updating: false,
    }

    // upon successful deletion ... 
    onDeleteSuccess = (e, user) => {
        this.setState({ 
            toRequests:true, // set true for redirecting 
        }) 
    }

    // upon successful request update, set updating to false
    onUpdateSuccess = ()=>{
        this.setState({
            updating:false,
        })
    }

    /* will be passed to CommentForm component as props.
    upon successfully adding comment, will set state commentAdded to true, and will make request for current request 
    and its comments, including newly added comment*/
    handleSubmit=()=>{
        this.setState({
            commentAdded: true
        })

        RequestApiService.getRequestById(this.props.match.params.id)
            .then(this.context.setCurrentRequest)
            .catch(this.context.setError)

        RequestApiService.getCommentsByRequestId(this.props.match.params.id)
            .then(this.context.setCurrentComments)
            .catch(this.context.setError)


    }

    componentWillMount=()=>{
        RequestApiService.getRequestById(this.props.match.params.id)
            .then(this.context.setCurrentRequest)
            .catch(this.context.setError)

        RequestApiService.getCommentsByRequestId(this.props.match.params.id)
            .then(this.context.setCurrentComments)
            .catch(this.context.setError)
        
    }

    renderComments=(comments)=>{
        let results
        if(comments.length===1){
            results = '1 Recommendation'
        }
        else{
            results = `${comments.length} Recommendations`
        }
        
        let commentItems = comments.map(comment=>
                <CommentItem key={comment.id} comment={comment}/>
            )

        return(
            <>
            <p className='commentsCount'>{results}</p>
            <div>{commentItems}</div>
            </>
        )
    }

    //when delete request button is clicked, will make DELETE request, GET new request list
    deleteRequest=()=>{
        RequestApiService.deleteRequest(this.context.currentRequest.id)
            .then(res=>{
                RequestApiService.getAllRequests()
                    .then(res=>{
                        this.context.setRequestList(res)
                        //GET user's requests
                        RequestApiService.getRequestsByUserId(UserService.getUserToken())
                            .then(this.context.setUsersList)
                            .catch(this.context.setError)
                    })
                    .catch(this.context.setError)
                
                this.onDeleteSuccess()
            })
            .catch(this.context.setError)
    }

    // when update request button is clicked, will set updating state to true
    updateRequest=()=>{
        this.setState({
            updating:true
        })
    }

    //if currently logged in user has user_id attached to current request, will render delete and edit buttons
    renderDeleteEditButtons=()=>{
        if(this.context.currentRequest.user_id===UserService.getUserToken()){
            return (
                <div className='editAndDeleteBox'>
                    <button onClick={this.deleteRequest} className='deleteButton'>Delete</button>
                    <button onClick={this.updateRequest} className='editButton'>Edit</button>
                </div>
            )
        }
    }

    /* if this.state.updating is false, renders request.
    if this.state.updating is true, will render UpdateRequest component*/
    renderRequest=()=>{
        if(this.state.updating===false){
            
            let lastName = this.context.currentRequest.last_name
            let lastInitial = new String(lastName).charAt(0)

            return(
                <div className='requestInfo'>
                    <h2>{this.context.currentRequest.product}</h2>
                    <div className='requestData'>
                        <div className='iconContainer'>
                            <i className="fas fa-user"></i>
                            <i className="fas fa-clock"></i>
                            <i className="fas fa-store"></i>
                        </div>
                        <div className='infoContainer'>
                            <p>{this.context.currentRequest.first_name} {lastInitial}.</p>
                            <p>{new Date(this.context.currentRequest.date).toLocaleDateString()}</p>
                            <p>{this.context.currentRequest.category}</p>
                        </div>
                    </div>
                    <p>{this.context.currentRequest.info}</p>
                    {this.renderDeleteEditButtons()}
                </div>
            )
        }
        else{
            return(
                <UpdateRequest request={this.context.currentRequest} onUpdateSuccess={this.onUpdateSuccess}/>
            )
        }
    }

    componentDidMount=()=>{
        this.context.clearError()
    }
    

    render(){
        // upon successful delete, redirect to your users requests
        if(this.state.toRequests===true){
            return <Redirect to='/requests/users'/>
        }
 
        return(
            <Error>
                <section className='indivRequestPage'>
                    <div className='sideBar'>
                    <button className='backButton' onClick={this.props.history.goBack}>Back</button>
                    </div>
                    <div className='mainBar'>
                    {this.renderRequest()}
                    <div className='commentSection'>
                        <CommentForm onSubmit={this.handleSubmit}/>
                        {this.renderComments(this.context.currentComments)}
                    </div>
                    </div>
                </section>
            </Error>
        )
    }
}