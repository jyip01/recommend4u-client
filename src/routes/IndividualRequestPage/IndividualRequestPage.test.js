import React from 'react'
import ReactDOM from 'react-dom'
import IndividualRequestPage from './IndividualRequestPage'
import {BrowserRouter} from 'react-router-dom'
import renderer from 'react-test-renderer'
import { createMemoryHistory } from 'history'


let match = {path: "/requests/:id", 
            url: "/requests/8", 
            isExact: true, 
            params: {id: "8"}
            }

it('renders without crashing',()=>{
    const history = createMemoryHistory('/requests/your')
    const div = document.createElement('div')
    ReactDOM.render(<BrowserRouter><IndividualRequestPage match={match} history={history}/></BrowserRouter>,div)
    ReactDOM.unmountComponentAtNode(div)
})

it('should match with snapshot', () => {
    const history = createMemoryHistory('/requests/your')
    const tree = renderer
      .create(
        <BrowserRouter>
          <IndividualRequestPage match={match} history={history}/>
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot()
})
