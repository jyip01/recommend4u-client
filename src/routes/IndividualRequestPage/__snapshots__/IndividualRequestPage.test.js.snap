// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match with snapshot 1`] = `
<section
  className="indivRequestPage"
>
  <div
    className="sideBar"
  >
    <button
      className="backButton"
      onClick={[Function]}
    >
      Back
    </button>
  </div>
  <div
    className="mainBar"
  >
    <div
      className="requestInfo"
    >
      <h2 />
      <div
        className="requestData"
      >
        <div
          className="iconContainer"
        >
          <i
            className="fas fa-user"
          />
          <i
            className="fas fa-clock"
          />
          <i
            className="fas fa-store"
          />
        </div>
        <div
          className="infoContainer"
        >
          <p>
             
            u
            .
          </p>
          <p>
            Invalid Date
          </p>
          <p />
        </div>
      </div>
      <p />
    </div>
    <div
      className="commentSection"
    >
      <div
        className="comment commentForm"
      >
        <p>
          <a
            href="/login"
            onClick={[Function]}
          >
            Login
          </a>
           to make a recommendation.
        </p>
      </div>
      <p
        className="commentsCount"
      >
        0 Recommendations
      </p>
      <div />
    </div>
  </div>
</section>
`;
